nodes/analyzer/length.py

```diff
@@ -0,0 +1,128 @@
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program; if not, write to the Free Software Foundation,
+#  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# ##### END GPL LICENSE BLOCK #####
+
+
+import bpy
+from bpy.props import BoolProperty
+from sverchok.node_tree import SverchCustomTreeNode
+from sverchok.data_structure import updateNode, match_long_repeat
+import numpy as np
+
+
+def edges_aux(vertices):
+    '''create auxiliary edges array '''
+    v_len = [len(v) for v in vertices]
+    v_len_max = max(v_len)
+    np_in = np.arange(v_len_max - 1)
+    np_edges = np.array([np_in, np_in + 1]).T
+
+    return [np_edges]
+
+
+def edges_length(meshes, gates, result):
+    '''calculate edges length '''
+
+    for vertices, edges in zip(*meshes):
+        np_verts = np.array(vertices)
+        print(np_verts.shape)
+        if type(edges[0]) in (list, tuple):
+            np_edges = np.array(edges)
+        else:
+            np_edges = edges[:len(vertices)-1, :]
+
+        vect = np_verts[np_edges[:, 0], :] - np_verts[np_edges[:, 1], :]
+        length = np.linalg.norm(vect, axis=1)
+        if not gates[1]:
+            length = np.sum(length)[np.newaxis]
+
+        result.append(length if gates[0] else length.tolist())
+
+    return result
+
+
+class SvLengthNode(bpy.types.Node, SverchCustomTreeNode):
+    '''
+    Triggers: Path / Edges length
+    Tooltip: Deformation between to states, edge elong a area variation
+    '''
+    bl_idname = 'SvLengthNode'
+    bl_label = 'Length'
+    bl_icon = 'MOD_SIMPLEDEFORM'
+
+    output_numpy = BoolProperty(
+        name='Output NumPy', description='output NumPy arrays',
+        default=False, update=updateNode)
+
+    sum_lengths = BoolProperty(
+        name='by Edge', description='individual lengths or the sum of them',
+        default=True, update=updateNode)
+
+    def draw_buttons(self, context, layout):
+        '''draw buttons on the Node'''
+        layout.prop(self, "sum_lengths", toggle=False)
+
+    def draw_buttons_ext(self, context, layout):
+        '''draw buttons on the N-panel'''
+        self.draw_buttons(context, layout)
+        layout.prop(self, "output_numpy", toggle=False)
+
+    def sv_init(self, context):
+        '''create sockets'''
+        sinw = self.inputs.new
+        sonw = self.outputs.new
+        sinw('VerticesSocket', "Verts")
+        sinw('StringsSocket', "Edges")
+
+        sonw('StringsSocket', "Length")
+
+    def get_data(self):
+        '''get all data from sockets'''
+        si = self.inputs
+        vertices = si['Verts'].sv_get(default=[[]])
+        edges_in = si['Edges'].sv_get(default=[[]])
+        if len(edges_in[0]) < 1:
+            edges_in = edges_aux(vertices)
```

Comment:
You could use the get_edge_list method to speed up linear edge generation :)

edges_in = get_edge_list(len(vertices))

and even (only read the input if connected) ..

if si[“Edges”].is_linked:
   edges = si['Edges'].sv_get(default=[[]])
else:
   edges = get_edge_list(len(vertices))

Reply:
Surprisingly after some tests the actual implementation (creating the edges in Numpy) is faster than using get_edge_list and the turn the result inside one array...
Maybe if the cache was already in numpy it would be faster but i'm not sure it worth to add it to the data_structure because is already pretty fast

Reply:
hm.. that is surprising indeed.. as the caching should be faster than any conversion.  Did you call update_edge_cache with a max number of edges you need to get from the cache? For a vectorized case when you have multiple lists of edges you can update the cache size once with the size of the largest list, then the cache will simply return the slice of an array which should be faster than generating arrays. You can even augment the cache to an arbitrary size (e.g. 1000 edges) so any calls to get_edge_list will never augment the cache.

Reply:
I did as you explain, calling the function once per update calling for the maximum length needed, and is still slower... maybe NumPy does some kind of memory optimisations under the hood... 🏹